flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java

```diff
@@ -351,21 +351,21 @@ public static TaskExecutor startTaskManager(
 
 		LOG.info("Starting TaskManager with ResourceID: {}", resourceID);
 
-		InetAddress remoteAddress = InetAddress.getByName(rpcService.getAddress());
+		InetAddress externalAddress = InetAddress.getByName(rpcService.getAddress());
```

Comment:
True. I'll try to fix the problem and update the e2e test case.

Reply:
After looking more into this, I'm not entirely sure whether we should support unresolvable external addresses.

The prerequisites are unclear to me. If an external address cannot be resolved by the TaskManager itself, then where can it be resolved? Do we assume all other components (JobManager and other TaskManagers) except for itself can resolve the external address? That sounds like a strong assumption to me, which may not always be true. E.g., deploying a Flink cluster over machines from two different sub-network, each sub-network has multiple machines/TMs. TMs in the same sub-network might not be able to resolve each other's external address. Another example, multiple TMs on the same machine.

I don't see a good way to cover all such cases. If we can only support the special case where external addresses can be resolved by all other components, I'm not sure whether we should do it given the efforts needed. Alternatively, user might use IP addresses to workaround in such cases.

Reply:
I thought that this is actually the problem we are trying to solve with this PR. Assume the `TM` is running in a Docker container which has an external name which is not resolvable from within the container. If all other Flink components can resolve this name, then it would work if the `TM` itself would not try to resolve the address.

At the moment we have the requirement that all `TM` components can resolve their own and the addresses of the other `TMs`. I think this requirement is too strict. What we actually need is that for a given `TM` `t` all other `TMs` `t'` can resolve the address of `t`. 

Of course, this means that if the user deploys multiple `TMs` into the same network where they cannot resolve each others external names, then Flink would not work. But this is ok because it violates the requirement that other `TMs` need to be able to resolve a `TM's` address. It is then the user's responsibility to configure the cluster accordingly.

Moreover, one could still enforce the current model by configuring the `bind-host` to the external address. If this is the case, then a `TM` would only start if it can resolve its own address.

Reply:
Alright, thanks for the explanation. I'll try to fix it.

Reply:
@tillrohrmann @xintongsong 
FYI
I think we do not have the requirement that other `TMs` need to be able to resolve a `TM's ` address. For current standalone Flink cluster on K8s, all the `TMs` could not resolve other's hostname. However, it works pretty well. This is the case for current master code without external address.

When  introducing the external address, i think it is same that we do not require other `TMs` need to be able to resolve a `TM's ` address.

Correct me if i am wrong somewhere.

Reply:
I think it currently works because the TM will resolve its own hostname and communicate the resolved IP address to the other TMs. If the TM cannot resolve it, then it will fail. If we introduce an external hostname, then this will no longer be possible.

Reply:
I agree with you that if the `TM` could not resolve its own address, it will fail. Indeed, it is problem we need to fix in this PR. However, i am just wondering given that we have fix this and do not require the `TM` could resolve its own external hostname, do we still need other `TMs` could resolve the external hostname? 

In my opinion every `TM` registers to the `JM` with hostname and ip. So all the `TMs` could get other `TM` hostname and ip. I am not sure about this.

Reply:
The problem with the external address is that it might resolve to an IP which is not known to the `TM`. This, of course, could also point towards a broken setup but if people want to run a `TM` behind a NAT, this could be a problem.